src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs

```diff
@@ -39,20 +34,50 @@ internal sealed class RegexFCD
         internal const int Boundary = 0x0040;
         internal const int ECMABoundary = 0x0080;
 
-        /*
-         * This is the one of the only two functions that should be called from outside.
-         * It takes a RegexTree and computes the set of chars that can start it.
-         */
-        internal static RegexPrefix FirstChars(RegexTree t)
+        private readonly ValueListBuilder<RegexFC> _fcStack;
+        private readonly ValueListBuilder<int> _intStack;
+        private bool _skipAllChildren;      // don't process any more children at the current level
+        private bool _skipchild;            // don't process the current child.
+        private bool _failed;
+
+        private RegexFCD(ValueListBuilder<RegexFC> fcStack, ValueListBuilder<int> intStack)
         {
-            RegexFCD s = new RegexFCD();
-            RegexFC fc = s.RegexFCFromRegexTree(t);
+            _fcStack = fcStack;
+            _intStack = intStack;
+            _failed = false;
+            _skipchild = false;
+            _skipAllChildren = false;
+        }
 
-            if (fc == null || fc._nullable)
+        /// <summary>
+        /// This is the one of the only two functions that should be called from outside.
+        /// It takes a RegexTree and computes the set of chars that can start it.
+        /// </summary>
+        internal static RegexPrefix? FirstChars(RegexTree t)
+        {
+            // Create/rent buffers
+            Span<int> intSpan = stackalloc int[StackBufferSize];
+            var intStack = new ValueListBuilder<int>(intSpan);
+            RegexFC[] fcSpan = ArrayPool<RegexFC>.Shared.Rent(StackBufferSize);
```

Comment:
We're renting an array of StackBufferSize?  I'd have thought we'd start with stack space of StackBufferSize.  I guess we can't as these aren't primitives?

Reply:
correct. it's a managed type which can't be stackalloced. I don't know why these restriction apply even to ref structs that always live on the stack. maybe @AndyAyersMS can answer?

Reply:
Right, stackallocation works for primitive/blitable types only.

Reply:
Stackalloc storage can't contain gc refs. Among other reasons, there is no mechanism for gc reporting from stackalloc buffers.

On the surface, the idea of allowing GC refs in a stackalloc buffer seems viable, if we can get sufficient safety guarantees. Say if the stackalloc is zero-init and immediately span wrapped and there is no safe way to view it as raw memory. With the right implementation (say a jit intrisnic Span ctor) the jit could know the type of the elements and create a new kind of gc info report for it. But we'd have to look into it in more detail to see if other problems crop up.

Reply:
Sounds interesting. Andy could you please open an issue for that? I think you could describe the matter far better than me.

Reply:
https://github.com/dotnet/csharplang/blob/master/proposals/fixed-sized-buffers.md and related issues are pretty much this. @VSadov has been working on this recently.

Reply:
There is a proposal and a prototype where you could just say
```C#
RegexFC[StackBufferSize] fcSpan = default; 

... just use fcSpan  as an indexable array-like struct ...

```

---
In reply to: [170391575](https://github.com/dotnet/corefx/pull/27314#discussion_r170391575) [](ancestors = 170391575)

Reply:
It could be useful in scenarios like this, but for now, I guess, array renting is a solution.

---
In reply to: [170393377](https://github.com/dotnet/corefx/pull/27314#discussion_r170393377) [](ancestors = 170393377,170391575)

Reply:
Supporting GC refs in span-wrapped stackallocs is similar to the fixed-sized buffer proposal but also seems slightly different, in perhaps interesting ways (?). It would allow for fixed or variable sized buffers and would be stack only, instead of supporting only fixed sized buffers but being storage agnostic. Would not require any language changes, just a new magic span constructor that the jit turns into stackalloc under the covers...

Reply:
Yes. The fixed size arrays are a syntactical sugar for ordinary struct types of appropriate size. You can have locals, parameters, fields of those types. They can even nest.
GC-tracked stackalloc allows for variable number of elements, but values must be stack-only. That would be basically the tradeoff.

---
In reply to: [170400977](https://github.com/dotnet/corefx/pull/27314#discussion_r170400977) [](ancestors = 170400977)